Tests BackupPrefsFragmentTest fail on emulator

On Pixel 9 API 34 BackupPrefsFragmentTest fail with log:

```
05-17 16:29:08.166 16281 16304 I TestRunner: started: verifyExportFile(com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest)
05-17 16:29:08.173 16281 16304 I UiAutomation: Initialized for user 0 on display 0
05-17 16:29:08.231  1510  2385 I AiAiEcho: Settings changed for uri: content://settings/secure/accessibility_enabled
05-17 16:29:08.244  1510  1787 I AiAiEcho: Predicting[0]:
05-17 16:29:08.245  1510  1787 I AiAiEcho: EchoTargets:
05-17 16:29:08.245  1510  1787 I AiAiEcho: Filtered by AiAi flag check:
05-17 16:29:08.245  1510  1787 I AiAiEcho: Ranked targets strategy: SORT, count: 0, ranking metadata:
05-17 16:29:08.251  1510  1787 I AiAiEcho: Predicting[0]:
05-17 16:29:08.251  1510  1787 I AiAiEcho: EchoTargets:
05-17 16:29:08.251  1510  1787 I AiAiEcho: Filtered by AiAi flag check:
05-17 16:29:08.251  1510  1787 I AiAiEcho: Ranked targets strategy: SORT, count: 0, ranking metadata:
05-17 16:29:08.251  1510  1787 I AiAiEcho: Predicting[0]:
05-17 16:29:08.251  1510  1787 I AiAiEcho: EchoTargets:
05-17 16:29:08.251  1510  1787 I AiAiEcho: Filtered by AiAi flag check:
05-17 16:29:08.251  1510  1787 I AiAiEcho: Ranked targets strategy: SORT, count: 0, ranking metadata:
05-17 16:29:08.253  1510  1787 I AiAiEcho: #postPredictionTargets: Sending updates to UISurface home with targets# 0 (types=[])
05-17 16:29:08.254  1510  1787 I AiAiEcho: #postPredictionTargets: Sending updates to UISurface media_data_manager with targets# 0 (types=[])
05-17 16:29:08.255  1510  1787 I AiAiEcho: #postPredictionTargets: Sending updates to UISurface lockscreen with targets# 0 (types=[])
05-17 16:29:08.257 16281 16304 I UiAutomationPermGranter: Attempting to grant android.permission.WRITE_EXTERNAL_STORAGE to com.amaze.filemanager.debug
05-17 16:29:08.267 16281 16304 I UiAutomationPermGranter: Attempting to grant android.permission.READ_EXTERNAL_STORAGE to com.amaze.filemanager.debug
05-17 16:29:08.343  1365  1365 D SsMediaDataProvider: Forwarding Smartspace updates []
05-17 16:29:08.343  1365  1365 W PageIndicator: Total number of pages invalid: 0. Assuming 1 page.
05-17 16:29:08.360  1494 15761 D MediaGrants: Removed 0 media_grants for 0 user for com.amaze.filemanager.debug. Reason: Mode changed: android:read_external_storage
05-17 16:29:08.368 16281 16304 W Settings: Setting always_finish_activities has moved from android.provider.Settings.System to android.provider.Settings.Global, returning read-only value.
05-17 16:29:08.377  9586  9651 I alns    : NGA disabled due to not being eligible.
05-17 16:29:08.385  1494 10872 D MediaGrants: Removed 0 media_grants for 0 user for com.amaze.filemanager.debug. Reason: Mode changed: android:read_external_storage
05-17 16:29:08.481   776   776 I wpa_supplicant: wlan0: CTRL-EVENT-BEACON-LOSS
05-17 16:29:08.505   400   400 I BpBinder: onLastStrongRef automatically unlinking death recipients:
05-17 16:29:08.578  1365  1401 E OpenGLRenderer: Unable to match the desired swap behavior.
05-17 16:29:08.596  9586  9586 I cndj    : onPause
05-17 16:29:08.610 16281 16281 D LifecycleMonitor: Lifecycle status change: com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2 in: PRE_ON_CREATE
05-17 16:29:08.611 16281 16281 V ActivityScenario: Activity lifecycle changed event received but ignored because the reported transition was not ON_CREATE while the last known transition was PRE_ON_CREATE
05-17 16:29:08.646  1055  1754 D EGL_emulation: app_time_stats: avg=6715.82ms min=6715.82ms max=6715.82ms count=1
05-17 16:29:08.663 16281 16326 D libEGL  : loaded /vendor/lib64/egl/libEGL_emulation.so
05-17 16:29:08.682 16281 16326 D libEGL  : loaded /vendor/lib64/egl/libGLESv1_CM_emulation.so
05-17 16:29:08.685 16281 16326 D libEGL  : loaded /vendor/lib64/egl/libGLESv2_emulation.so
05-17 16:29:08.752   564  6041 W OpenGLRenderer: Failed to choose config with EGL_SWAP_BEHAVIOR_PRESERVED, retrying without...
05-17 16:29:08.754   564  6041 W OpenGLRenderer: Failed to initialize 101010-2 format, error = EGL_SUCCESS
05-17 16:29:08.770 16281 16281 W lemanager.debug: Accessing hidden method Landroid/view/View;->computeFitSystemWindows(Landroid/graphics/Rect;Landroid/graphics/Rect;)Z (unsupported, reflection, allowed)
05-17 16:29:08.771 16281 16281 W lemanager.debug: Accessing hidden method Landroid/view/ViewGroup;->makeOptionalFitsSystemWindows()V (unsupported, reflection, allowed)
05-17 16:29:08.775 16281 16281 I AppCompatViewInflater: app:theme is now deprecated. Please move to using android:theme instead.
05-17 16:29:08.810 16281 16281 D CompatibilityChangeReporter: Compat change id reported: 210923482; UID 10190; state: ENABLED
05-17 16:29:08.839 16281 16281 D LifecycleMonitor: Lifecycle status change: com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2 in: CREATED
05-17 16:29:08.840 16281 16281 V ActivityScenario: Update currentActivityStage to CREATED, currentActivity=com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2
05-17 16:29:08.858  9586 12502 V ClearcutMetricXmitter: Transmission is done.
05-17 16:29:08.881 16281 16281 D LifecycleMonitor: Lifecycle status change: com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2 in: STARTED
05-17 16:29:08.881 16281 16281 V ActivityScenario: Update currentActivityStage to STARTED, currentActivity=com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2
05-17 16:29:08.884 16281 16281 D LifecycleMonitor: Lifecycle status change: com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2 in: RESUMED
05-17 16:29:08.884 16281 16281 V ActivityScenario: Update currentActivityStage to RESUMED, currentActivity=com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2
05-17 16:29:08.889 16281 16281 D CompatibilityChangeReporter: Compat change id reported: 237531167; UID 10190; state: DISABLED
05-17 16:29:08.891 16281 16281 W lemanager.debug: Accessing hidden field Lcom/android/internal/policy/DecorView;->mWindow:Lcom/android/internal/policy/PhoneWindow; (unsupported, reflection, allowed)
05-17 16:29:08.893 16281 16281 W OpenGLRenderer: Unknown dataspace 0
05-17 16:29:08.948 16281 16325 W OpenGLRenderer: Failed to choose config with EGL_SWAP_BEHAVIOR_PRESERVED, retrying without...
05-17 16:29:08.949 16281 16325 W OpenGLRenderer: Failed to initialize 101010-2 format, error = EGL_SUCCESS
05-17 16:29:08.975   139   139 I hwservicemanager: getTransport: Cannot find entry android.hardware.graphics.mapper@4.0::IMapper/default in either framework or device VINTF manifest.
05-17 16:29:08.975 16281 16325 I Gralloc4: mapper 4.x is not supported
05-17 16:29:09.051 16281 16325 E OpenGLRenderer: Unable to match the desired swap behavior.
05-17 16:29:09.222   564   592 W ziparchive: Unable to open '/data/app/~~HmbM_NR-0wKvBIWPBlOyYw==/com.amaze.filemanager.debug-rk_0oRRSR4lPIAzLS_fKtA==/base.dm': No such file or directory
05-17 16:29:09.230   564   592 I ActivityTaskManager: Displayed com.amaze.filemanager.debug/com.amaze.filemanager.ui.activities.PreferencesActivity for user 0: +828ms
05-17 16:29:09.248  1365  1394 W WindowOnBackDispatcher: sendCancelIfRunning: isInProgress=falsecallback=android.view.ViewRootImpl$$ExternalSyntheticLambda17@ea75c48
05-17 16:29:09.251   564  2537 W InputManager-JNI: Input channel object 'ce8914d Splash Screen com.amaze.filemanager.debug (client)' was disposed without first being removed with the input manager!
05-17 16:29:09.299  1055  1108 I ImeTracker: com.google.android.apps.nexuslauncher:71334dab: onRequestHide at ORIGIN_CLIENT_HIDE_SOFT_INPUT reason HIDE_SOFT_INPUT
05-17 16:29:09.301   564  2516 I ImeTracker: com.google.android.apps.nexuslauncher:71334dab: onCancelled at PHASE_SERVER_SHOULD_HIDE
05-17 16:29:09.310  9586  9586 I cndj    : onStop
05-17 16:29:09.313   564   582 I ImeTracker: com.amaze.filemanager.debug:665bf552: onRequestHide at ORIGIN_SERVER_HIDE_INPUT reason HIDE_UNSPECIFIED_WINDOW
05-17 16:29:09.313   564   582 I ImeTracker: com.amaze.filemanager.debug:665bf552: onCancelled at PHASE_SERVER_SHOULD_HIDE
05-17 16:29:09.325  1055  1885 D OneSearchSuggestProvider: Shut down the binder channel
05-17 16:29:09.339  1365  1401 D EGL_emulation: app_time_stats: avg=203462.23ms min=203462.23ms max=203462.23ms count=1
05-17 16:29:09.348  1055  1776 I IPCThreadState: oneway function results for code 2 on binder at 0x78dcd4688270 will be dropped but finished with status UNKNOWN_TRANSACTION
05-17 16:29:09.387  1637  1637 I GoogleInputMethodService: GoogleInputMethodService.onFinishInput():2043
05-17 16:29:09.389  1637  1637 I Module  : DeviceLockedStatusModuleProvider$Module.updateDeviceLockedStatus():100 repeatCheckTimes = 0, locked = false
05-17 16:29:09.392  1637  1637 I GoogleInputMethodService: GoogleInputMethodService.onStartInput():1293 onStartInput(EditorInfo{EditorInfo{packageName=com.amaze.filemanager.debug, inputType=0, inputTypeString=NULL, enableLearning=false, autoCorrection=false, autoComplete=false, imeOptions=0, privateImeOptions=null, actionName=UNSPECIFIED, actionLabel=null, initialSelStart=-1, initialSelEnd=-1, initialCapsMode=0, label=null, fieldId=0, fieldName=null, extras=null, hintText=null, hintLocales=[]}}, false)
05-17 16:29:09.393  1637  1637 I Module  : DeviceLockedStatusModuleProvider$Module.updateDeviceLockedStatus():100 repeatCheckTimes = 1, locked = false
05-17 16:29:09.419  1256  2118 I FontLog : (REDACTED) Received query %s, URI %s
05-17 16:29:09.422  1256  2118 I FontLog : (REDACTED) Query [%s] resolved to %s
05-17 16:29:09.439  1256  2118 I FontLog : (REDACTED) Fetch %s end status %s
05-17 16:29:09.453 16281 16281 D LifecycleMonitor: Lifecycle status change: com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2 in: PAUSED
05-17 16:29:09.457 16281 16281 V ActivityScenario: Update currentActivityStage to PAUSED, currentActivity=com.amaze.filemanager.ui.activities.PreferencesActivity@fcc2ab2
05-17 16:29:09.466  1256  2118 I FontLog : (REDACTED) Pulling font file for id = %d, cache size = %d
05-17 16:29:09.472   400   400 I BpBinder: onLastStrongRef automatically unlinking death recipients:
05-17 16:29:09.484   334   334 D Zygote  : Forked child process 16339
05-17 16:29:09.486  1256  2118 I FontLog : (REDACTED) Pulling font file for id = %d, cache size = %d
05-17 16:29:09.503 16339 16339 I ager.debug.test: Late-enabling -Xcheck:jni
05-17 16:29:09.531 16339 16339 I ager.debug.test: Using CollectorTypeCC GC.
05-17 16:29:09.532 16339 16339 W ager.debug.test: Unexpected CPU variant for x86: x86_64.
05-17 16:29:09.532 16339 16339 W ager.debug.test: Known variants: atom, sandybridge, silvermont, goldmont, goldmont-plus, tremont, kabylake, default
05-17 16:29:09.537   413   428 I adbd    : jdwp connection from 16339
05-17 16:29:09.592 16339 16339 D CompatibilityChangeReporter: Compat change id reported: 171979766; UID 10203; state: ENABLED
05-17 16:29:09.593 16339 16339 D CompatibilityChangeReporter: Compat change id reported: 242716250; UID 10203; state: ENABLED
05-17 16:29:09.609 16339 16339 D ApplicationLoaders: Returning zygote-cached class loader: /system/framework/android.test.base.jar
05-17 16:29:09.616 16339 16339 D nativeloader: Configuring clns-6 for other apk /system/framework/android.test.mock.jar. target_sdk_version=34, uses_libraries=ALL, library_path=/data/app/~~IveKVUAbWWJFEyI9FJuvYA==/com.amaze.filemanager.debug.test-O1qdUTdQng6dpph9Iy63sA==/lib/x86_64, permitted_path=/data:/mnt/expand:/data/user/0/com.amaze.filemanager.debug.test
05-17 16:29:09.618 16339 16339 D ApplicationLoaders: Returning zygote-cached class loader: /system/framework/android.test.base.jar
05-17 16:29:09.623 16339 16339 D nativeloader: Configuring clns-7 for other apk /system/framework/android.test.runner.jar. target_sdk_version=34, uses_libraries=ALL, library_path=/data/app/~~IveKVUAbWWJFEyI9FJuvYA==/com.amaze.filemanager.debug.test-O1qdUTdQng6dpph9Iy63sA==/lib/x86_64, permitted_path=/data:/mnt/expand:/data/user/0/com.amaze.filemanager.debug.test
05-17 16:29:09.627 16339 16339 W ziparchive: Unable to open '/data/app/~~IveKVUAbWWJFEyI9FJuvYA==/com.amaze.filemanager.debug.test-O1qdUTdQng6dpph9Iy63sA==/base.dm': No such file or directory
05-17 16:29:09.632 16339 16339 W ziparchive: Unable to open '/data/app/~~IveKVUAbWWJFEyI9FJuvYA==/com.amaze.filemanager.debug.test-O1qdUTdQng6dpph9Iy63sA==/base.dm': No such file or directory
05-17 16:29:09.661 16339 16339 D nativeloader: Configuring clns-8 for other apk /data/app/~~IveKVUAbWWJFEyI9FJuvYA==/com.amaze.filemanager.debug.test-O1qdUTdQng6dpph9Iy63sA==/base.apk. target_sdk_version=34, uses_libraries=, library_path=/data/app/~~IveKVUAbWWJFEyI9FJuvYA==/com.amaze.filemanager.debug.test-O1qdUTdQng6dpph9Iy63sA==/lib/x86_64, permitted_path=/data:/mnt/expand:/data/user/0/com.amaze.filemanager.debug.test
05-17 16:29:09.679 16339 16339 V GraphicsEnvironment: Currently set values for:
05-17 16:29:09.679 16339 16339 V GraphicsEnvironment:   angle_gl_driver_selection_pkgs=[]
05-17 16:29:09.679 16339 16339 V GraphicsEnvironment:   angle_gl_driver_selection_values=[]
05-17 16:29:09.680 16339 16339 V GraphicsEnvironment: ANGLE GameManagerService for com.amaze.filemanager.debug.test: false
05-17 16:29:09.680 16339 16339 V GraphicsEnvironment: com.amaze.filemanager.debug.test is not listed in per-application setting
05-17 16:29:09.681 16339 16339 V GraphicsEnvironment: Neither updatable production driver nor prerelease driver is supported.
05-17 16:29:09.707 16339 16355 D libEGL  : loaded /vendor/lib64/egl/libEGL_emulation.so
05-17 16:29:09.736 16339 16355 D libEGL  : loaded /vendor/lib64/egl/libGLESv1_CM_emulation.so
05-17 16:29:09.739 16339 16339 D CompatibilityChangeReporter: Compat change id reported: 237531167; UID 10203; state: DISABLED
05-17 16:29:09.741 16339 16339 W OpenGLRenderer: Unknown dataspace 0
05-17 16:29:09.751 16339 16355 D libEGL  : loaded /vendor/lib64/egl/libGLESv2_emulation.so
05-17 16:29:09.770   400   540 W ServiceManagerCppClient: Permission failure: android.permission.ROTATE_SURFACE_FLINGER from uid=10203 pid=0
05-17 16:29:09.770   400   540 D PermissionCache: checking android.permission.ROTATE_SURFACE_FLINGER for uid=10203 => denied (281 us)
05-17 16:29:09.771   400   540 W ServiceManagerCppClient: Permission failure: android.permission.INTERNAL_SYSTEM_WINDOW from uid=10203 pid=0
05-17 16:29:09.771   400   540 D PermissionCache: checking android.permission.INTERNAL_SYSTEM_WINDOW for uid=10203 => denied (464 us)
05-17 16:29:09.771   400   540 W ServiceManagerCppClient: Permission failure: android.permission.ACCESS_SURFACE_FLINGER from uid=10203 pid=0
05-17 16:29:09.771   400   540 D PermissionCache: checking android.permission.ACCESS_SURFACE_FLINGER for uid=10203 => denied (282 us)
05-17 16:29:09.774 16281 16304 E TestRunner: failed: verifyExportFile(com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest)
05-17 16:29:09.777 16281 16304 E TestRunner: ----- begin exception -----
05-17 16:29:09.777 16281 16304 E TestRunner: java.lang.RuntimeException: java.io.FileNotFoundException: /storage/emulated/0/amaze_backup.json: open failed: ENOENT (No such file or directory)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.runner.MonitoringInstrumentation.runOnMainSync(MonitoringInstrumentation.java:497)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.core.app.ActivityScenario.onActivity(ActivityScenario.java:807)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest.verifyExportFile(BackupPrefsFragmentTest.kt:123)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at java.lang.reflect.Method.invoke(Native Method)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:150)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:162)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.Suite.runChild(Suite.java:128)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.Suite.runChild(Suite.java:27)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:68)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:59)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:463)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2402)
05-17 16:29:09.777 16281 16304 E TestRunner: Caused by: java.io.FileNotFoundException: /storage/emulated/0/amaze_backup.json: open failed: ENOENT (No such file or directory)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at libcore.io.IoBridge.open(IoBridge.java:574)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at java.io.FileInputStream.<init>(FileInputStream.java:160)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest.verifyExportFile$lambda$2(BackupPrefsFragmentTest.kt:135)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest.$r8$lambda$gyyh-b5CTRMTC3G0mQ8tVq9KH9c(Unknown Source:0)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest$$ExternalSyntheticLambda1.perform(D8$$SyntheticClass:0)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.core.app.ActivityScenario.lambda$onActivity$2$androidx-test-core-app-ActivityScenario(ActivityScenario.java:794)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at androidx.test.core.app.ActivityScenario$$ExternalSyntheticLambda2.run(D8$$SyntheticClass:0)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:487)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at android.app.Instrumentation$SyncRunnable.run(Instrumentation.java:2420)
05-17 16:29:09.777 16281 16304 E TestRunner: 	at android.os.Handler.handleCallback(Handler.java:958)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at android.os.Handler.dispatchMessage(Handler.java:99)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at android.os.Looper.loopOnce(Looper.java:205)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at android.os.Looper.loop(Looper.java:294)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at android.app.ActivityThread.main(ActivityThread.java:8177)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at java.lang.reflect.Method.invoke(Native Method)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
05-17 16:29:09.779 16281 16304 E TestRunner: Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at libcore.io.Linux.open(Native Method)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at libcore.io.BlockGuardOs.open(BlockGuardOs.java:274)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:8063)
05-17 16:29:09.779 16281 16304 E TestRunner: 	at libcore.io.IoBridge.open(IoBridge.java:560)
05-17 16:29:09.779 16281 16304 E TestRunner: 	... 17 more
05-17 16:29:09.779 16281 16304 E TestRunner: ----- end exception -----
05-17 16:29:09.790 16281 16304 I TestRunner: finished: verifyExportFile(com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest)

java.lang.RuntimeException: java.io.FileNotFoundException: /storage/emulated/0/amaze_backup.json: open failed: ENOENT (No such file or directory)
at androidx.test.runner.MonitoringInstrumentation.runOnMainSync(MonitoringInstrumentation.java:497)
at androidx.test.core.app.ActivityScenario.onActivity(ActivityScenario.java:807)
at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest.verifyExportFile(BackupPrefsFragmentTest.kt:123)
... 33 trimmed
Caused by: java.io.FileNotFoundException: /storage/emulated/0/amaze_backup.json: open failed: ENOENT (No such file or directory)
at libcore.io.IoBridge.open(IoBridge.java:574)
at java.io.FileInputStream.<init>(FileInputStream.java:160)
at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest.verifyExportFile$lambda$2(BackupPrefsFragmentTest.kt:135)
at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest.$r8$lambda$gyyh-b5CTRMTC3G0mQ8tVq9KH9c(Unknown Source:0)
at com.amaze.filemanager.ui.fragments.BackupPrefsFragmentTest$$ExternalSyntheticLambda1.perform(D8$$SyntheticClass:0)
at androidx.test.core.app.ActivityScenario.lambda$onActivity$2$androidx-test-core-app-ActivityScenario(ActivityScenario.java:794)
at androidx.test.core.app.ActivityScenario$$ExternalSyntheticLambda2.run(D8$$SyntheticClass:0)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:487)
at java.util.concurrent.FutureTask.run(FutureTask.java:264)
at android.app.Instrumentation$SyncRunnable.run(Instrumentation.java:2420)
at android.os.Handler.handleCallback(Handler.java:958)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:205)
at android.os.Looper.loop(Looper.java:294)
at android.app.ActivityThread.main(ActivityThread.java:8177)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
at libcore.io.Linux.open(Native Method)
at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
at libcore.io.BlockGuardOs.open(BlockGuardOs.java:274)
at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:8063)
at libcore.io.IoBridge.open(IoBridge.java:560)
... 17 more
```

--- a/app/src/androidTest/java/com/amaze/filemanager/test/StoragePermissionHelper.kt
+++ b/app/src/androidTest/java/com/amaze/filemanager/test/StoragePermissionHelper.kt
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2014-2022 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
+ * Copyright (C) 2014-2025 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
  * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
  *
  * This file is part of Amaze File Manager.
@@ -20,50 +20,54 @@
 
 package com.amaze.filemanager.test
 
-// import android.content.Intent
-// import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
-// import android.net.Uri
-// import android.os.Build
-// import android.os.Build.VERSION.SDK_INT
-// import android.os.Environment
-// import android.provider.Settings
-// import android.widget.Switch
-// import androidx.test.platform.app.InstrumentationRegistry
-// import androidx.test.uiautomator.UiDevice
-// import androidx.test.uiautomator.UiSelector
-// import org.junit.Assert.assertTrue
+import android.content.Context
+import androidx.test.core.app.ActivityScenario
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.matcher.ViewMatchers.withText
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.uiautomator.By
+import androidx.test.uiautomator.UiDevice
+import androidx.test.uiautomator.UiSelector
+import com.amaze.filemanager.R
+import com.amaze.filemanager.ui.activities.MainActivity
 
 object StoragePermissionHelper {
     /**
-     * This method is intended for Android R or above devices to obtain MANAGE_EXTERNAL_STORAGE
-     * permission via UI Automator framework when running relevant Espresso tests.
-     *
-     * This method is flat commented out because UI Automator requires Android SDK 18, while
-     * currently we still want to support SDK 14.
+     * From https://github.com/android/android-test/issues/1658#issue-1551755250
+     * HACK this grants access to external storage "manually" because other solutions don't seem
+     * to set the permission.
      */
     @JvmStatic
-    fun obtainManageAppAllFileAccessPermissionAutomatically() {
-//        if (!Environment.isExternalStorageManager() && SDK_INT > Build.VERSION_CODES.R) {
-//            InstrumentationRegistry.getInstrumentation().run {
-//                val device = androidx.test.uiautomator.UiDevice.getInstance(this)
-//                val context = this.targetContext
-//                device.pressHome()
-//                val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
-//                    .setData(Uri.parse("package:${context.packageName}"))
-//                    .addFlags(FLAG_ACTIVITY_NEW_TASK)
-//                context.startActivity(intent)
-//                val switch = device.findObject(
-//                    androidx.test.uiautomator.UiSelector()
-//                        .packageName("com.android.settings")
-//                        .className(Switch::class.java.name)
-//                        .resourceId("android:id/switch_widget")
-//                )
-//                switch.click()
-//                assertTrue(switch.isChecked)
-//                device.pressHome()
-//            }
-//        }
-//        assertTrue(Environment.isExternalStorageManager())
-        return // Try to get codacy happy if they ever check me... pretend I am doing something
+    fun grantManageStoragePermission() {
+        // Ensure that an activity that has the dialog is launched
+        ActivityScenario.launch(MainActivity::class.java)
+
+        val context: Context = InstrumentationRegistry.getInstrumentation().targetContext
+        val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+
+        val amazeResources = context.packageManager.getResourcesForApplication(context.packageName)
+        val grantPermissionExplanation = amazeResources.getString(R.string.grant_all_files_permission)
+
+        if (device.hasObject(By.text(grantPermissionExplanation))) {
+            // First press Amaze's grant button
+            onView(withText(R.string.grant)).perform(click())
+
+            // Identifier names are taken here:
+            // https://cs.android.com/android/platform/superproject/+/master:packages/apps/Settings/res/values/strings.xml
+            val resources = context.packageManager.getResourcesForApplication("com.android.settings")
+            val resId =
+                resources.getIdentifier(
+                    "permit_manage_external_storage",
+                    "string",
+                    "com.android.settings",
+                )
+            val permitManageExternalStorage = resources.getString(resId)
+
+            val grantToggle =
+                device.findObject(UiSelector().textMatches("(?i)$permitManageExternalStorage"))
+            grantToggle.click()
+            device.pressBack()
+        }
     }
 }

--- a/app/src/androidTest/java/com/amaze/filemanager/ui/fragments/BackupPrefsFragmentTest.kt
+++ b/app/src/androidTest/java/com/amaze/filemanager/ui/fragments/BackupPrefsFragmentTest.kt
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2014-2022 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
+ * Copyright (C) 2014-2025 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
  * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
  *
  * This file is part of Amaze File Manager.
@@ -25,6 +25,8 @@ import android.content.Context
 import android.content.Intent
 import android.content.SharedPreferences
 import android.net.Uri
+import android.os.Build.VERSION.SDK_INT
+import android.os.Build.VERSION_CODES.TIRAMISU
 import androidx.lifecycle.Lifecycle
 import androidx.preference.PreferenceManager
 import androidx.test.core.app.ActivityScenario
@@ -36,13 +38,16 @@ import androidx.test.espresso.matcher.ViewMatchers.withText
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import androidx.test.rule.GrantPermissionRule
 import com.amaze.filemanager.R
+import com.amaze.filemanager.test.StoragePermissionHelper
 import com.amaze.filemanager.ui.activities.PreferencesActivity
 import com.amaze.filemanager.ui.fragments.preferencefragments.BackupPrefsFragment
 import com.google.gson.GsonBuilder
 import com.google.gson.reflect.TypeToken
-import org.junit.Assert
 import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
 import org.junit.Assert.assertTrue
+import org.junit.Assert.fail
+import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
@@ -59,23 +64,46 @@ class BackupPrefsFragmentTest {
         GrantPermissionRule
             .grant(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
 
-    /** Test exporting */
-    @Test
-    fun testExport() {
-        val backupPrefsFragment = BackupPrefsFragment()
+    @Rule
+    @JvmField
+    val notificationPermissionRule: GrantPermissionRule =
+        if (SDK_INT >= TIRAMISU) {
+            GrantPermissionRule.grant(android.Manifest.permission.POST_NOTIFICATIONS)
+        } else {
+            GrantPermissionRule.grant()
+        }
 
+    /**
+     * Storage permission is needed for saving the preferences to a user accessible file
+     */
+    @Before
+    fun grantManageStoragePermission() {
+        StoragePermissionHelper.grantManageStoragePermission()
+    }
+
+    /** Test exporting and reimporting preferences */
+    @Test
+    fun testPreferencesExportImport() {
+        val context = ApplicationProvider.getApplicationContext<Context>()
+
+        val exportFile = File("$storagePath${File.separator}$fileName")
+        exportFile.delete() // delete if already exists
+
+        export(context, exportFile)
+        import(exportFile)
+    }
+
+    /**
+     * Test whether the exported file contains the expected preference values
+     */
+    private fun export(
+        context: Context,
+        exportFile: File,
+    ) {
+        val backupPrefsFragment = BackupPrefsFragment()
         val activityScenario = ActivityScenario.launch(PreferencesActivity::class.java)
 
         activityScenario.moveToState(Lifecycle.State.STARTED)
-
-        val exportFile =
-            File(
-                storagePath +
-                    File.separator +
-                    fileName,
-            )
-
-        exportFile.delete() // delete if already exists
 
         activityScenario.onActivity {
             it.supportFragmentManager.beginTransaction()
@@ -85,54 +113,21 @@ class BackupPrefsFragmentTest {
             backupPrefsFragment.exportPrefs()
         }
 
-        val tempFile =
-            File(
-                ApplicationProvider.getApplicationContext<Context>().cacheDir.absolutePath +
-                    File.separator +
-                    fileName,
-            )
+        val tempFile = File("${context.cacheDir.absolutePath}${File.separator}$fileName")
 
-        Assert.assertTrue(tempFile.exists())
+        assertTrue(tempFile.exists())
 
-        // terrible hack :cringe:
         onView(withId(R.id.home)).perform(ViewActions.click())
-        Thread.sleep(500)
-
         onView(withText(R.string.save)).perform(ViewActions.click())
-        Thread.sleep(500)
 
-        Assert.assertTrue(exportFile.exists())
-    }
-
-    /** Test whether the exported file contains the expected preference values */
-    @Test
-    fun verifyExportFile() {
-        val backupPrefsFragment = BackupPrefsFragment()
-
-        val activityScenario = ActivityScenario.launch(PreferencesActivity::class.java)
-
-        activityScenario.moveToState(Lifecycle.State.STARTED)
-
-        val file =
-            File(
-                storagePath +
-                    File.separator +
-                    fileName,
-            )
+        assertTrue(exportFile.exists())
 
         activityScenario.onActivity { preferencesActivity ->
-            preferencesActivity.supportFragmentManager.beginTransaction()
-                .add(backupPrefsFragment, null)
-                .commitNow()
-
-            val preferences =
-                PreferenceManager
-                    .getDefaultSharedPreferences(preferencesActivity)
-
+            val preferences = PreferenceManager.getDefaultSharedPreferences(preferencesActivity)
             val preferenceMap: Map<String?, *> = preferences.all
 
             val inputString =
-                file
+                exportFile
                     .inputStream()
                     .bufferedReader()
                     .use {
@@ -150,34 +145,27 @@ class BackupPrefsFragmentTest {
                     )
 
             for ((key, value) in preferenceMap) {
-                var mapValue = importMap[key]
-
-                if (mapValue!!::class.simpleName.equals("Double")) {
-                    mapValue = (mapValue as Double).toInt() // since Gson parses Integer as Double
-                }
+                val importedValue = importMap[key]
+                val mapValue =
+                    if (importedValue != null && importedValue::class.simpleName.equals("Double")) {
+                        (importedValue as Double).toInt() // since Gson parses Integer as Double
+                    } else {
+                        importedValue
+                    }
 
                 assertEquals("Difference found at key $key", value, mapValue)
             }
         }
     }
 
-    /** Test import */
-    @Test
-    fun testImport() {
-        val backupPrefsFragment = BackupPrefsFragment()
-
+    /**
+     * Test whether the imported preferences contains the expected values
+     */
+    private fun import(exportFile: File) {
         val activityScenario = ActivityScenario.launch(PreferencesActivity::class.java)
-
         activityScenario.moveToState(Lifecycle.State.STARTED)
 
-        val exportFile =
-            File(
-                storagePath +
-                    File.separator +
-                    fileName,
-            )
-
-        exportFile.delete() // delete if already exists
+        val backupPrefsFragment = BackupPrefsFragment()
 
         activityScenario.onActivity { preferencesActivity ->
             preferencesActivity.supportFragmentManager.beginTransaction()
@@ -207,18 +195,19 @@ class BackupPrefsFragmentTest {
             val importMap: Map<String?, *> =
                 GsonBuilder()
                     .create()
-                    .fromJson(
-                        inputString,
-                        type,
-                    )
+                    .fromJson(inputString, type)
 
-            val preferences =
-                PreferenceManager
-                    .getDefaultSharedPreferences(preferencesActivity)
+            val preferences = PreferenceManager.getDefaultSharedPreferences(preferencesActivity)
 
             val preferenceMap: Map<String?, *> = preferences.all
 
-            for ((key, value) in preferenceMap) {
+            assertFalse(preferenceMap.containsKey(null))
+
+            for ((k, v) in preferenceMap) {
+                // This cast tells the kotlin type checker that fail() never returns
+                val key = k ?: (fail() as Nothing)
+                val value = v ?: (fail() as Nothing)
+
                 assertTrue("checkPrefEqual($key) failed", checkPrefEqual(preferences, importMap, key, value))
             }
         }
@@ -227,10 +216,10 @@ class BackupPrefsFragmentTest {
     private fun checkPrefEqual(
         preferences: SharedPreferences,
         importMap: Map<String?, *>,
-        key: String?,
-        value: Any?,
+        key: String,
+        value: Any,
     ): Boolean {
-        when (value!!::class.simpleName) {
+        when (value::class.simpleName) {
             "Boolean" -> return importMap[key] as Boolean ==
                 preferences.getBoolean(key, false)
             "Float" ->

--- a/app/src/androidTest/java/com/amaze/filemanager/ui/fragments/TabFragmentTest.kt
+++ b/app/src/androidTest/java/com/amaze/filemanager/ui/fragments/TabFragmentTest.kt
@@ -1,23 +1,18 @@
 package com.amaze.filemanager.ui.fragments
 
-import android.content.Context
 import android.content.pm.ActivityInfo
 import android.os.Build.VERSION.SDK_INT
 import android.os.Build.VERSION_CODES.TIRAMISU
 import androidx.test.espresso.Espresso.onView
-import androidx.test.espresso.action.ViewActions.click
 import androidx.test.espresso.action.ViewActions.swipeLeft
 import androidx.test.espresso.action.ViewActions.swipeRight
 import androidx.test.espresso.matcher.ViewMatchers.withId
-import androidx.test.espresso.matcher.ViewMatchers.withText
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import androidx.test.platform.app.InstrumentationRegistry
 import androidx.test.rule.ActivityTestRule
 import androidx.test.rule.GrantPermissionRule
-import androidx.test.uiautomator.By
-import androidx.test.uiautomator.UiDevice
-import androidx.test.uiautomator.UiSelector
 import com.amaze.filemanager.R
+import com.amaze.filemanager.test.StoragePermissionHelper
 import com.amaze.filemanager.ui.activities.MainActivity
 import org.junit.Before
 import org.junit.Rule
@@ -51,39 +46,9 @@ class TabFragmentTest {
             GrantPermissionRule.grant()
         }
 
-    /**
-     * From https://github.com/android/android-test/issues/1658#issue-1551755250
-     * HACK this grants access to external storage "manually" because other solutions don't seem
-     * to set the permission.
-     */
     @Before
     fun grantManageStoragePermission() {
-        val context: Context = InstrumentationRegistry.getInstrumentation().targetContext
-        val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
-
-        val amazeResources = context.packageManager.getResourcesForApplication(context.packageName)
-        val grantPermissionExplanation = amazeResources.getString(R.string.grant_all_files_permission)
-
-        if (device.hasObject(By.text(grantPermissionExplanation))) {
-            // First press Amaze's grant button
-            onView(withText(R.string.grant)).perform(click())
-
-            // Identifier names are taken here:
-            // https://cs.android.com/android/platform/superproject/+/master:packages/apps/Settings/res/values/strings.xml
-            val resources = context.packageManager.getResourcesForApplication("com.android.settings")
-            val resId =
-                resources.getIdentifier(
-                    "permit_manage_external_storage",
-                    "string",
-                    "com.android.settings",
-                )
-            val permitManageExternalStorage = resources.getString(resId)
-
-            val grantToggle =
-                device.findObject(UiSelector().textMatches("(?i)$permitManageExternalStorage"))
-            grantToggle.click()
-            device.pressBack()
-        }
+        StoragePermissionHelper.grantManageStoragePermission()
     }
 
     /**

